File: examples/sensor_viewer.cpp
#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdlib>
#include <igris_sdk/channel_factory.hpp>
#include <igris_sdk/igris_c_client.hpp>
#include <igris_sdk/igris_c_msgs.hpp>
#include <igris_sdk/subscriber.hpp>
#include <iomanip>
#include <iostream>
#include <mutex>
#include <numeric>
#include <opencv2/opencv.hpp>
#include <sstream>
#include <string>
#include <thread>
#include <vector>

struct status {
    std::string name;
    float fps           = 0.0f;
    float bytes_per_sec = 0.0f;
};

namespace {
int domain_id = 10;

using namespace igris_sdk;
using CompressedMessage = igris_c::msg::dds::CompressedMessage;

std::atomic<bool> g_stop{false};

void handle_signal(int) { g_stop.store(true); }

int getenv_int(const char *name, int fallback) {
    const char *value = std::getenv(name);
    if (!value || value[0] == '\0') {
        return fallback;
    }
    char *end   = nullptr;
    long parsed = std::strtol(value, &end, 10);
    if (end == value) {
        return fallback;
    }
    return static_cast<int>(parsed);
}

struct StreamStats {
    std::chrono::steady_clock::time_point last_time = std::chrono::steady_clock::now();
    double fps                                      = 0.0;
    double bytes_per_sec                            = 0.0;
};

void update_stats(StreamStats &stats, size_t bytes) {
    auto now     = std::chrono::steady_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(now - stats.last_time).count();
    if (elapsed > 0) {
        double seconds  = static_cast<double>(elapsed) / 1'000'000.0;
        double inst_fps = 1.0 / seconds;
        double inst_bps = static_cast<double>(bytes) / seconds;
        double alpha    = 0.15;
        if (stats.fps <= 0.0) {
            stats.fps           = inst_fps;
            stats.bytes_per_sec = inst_bps;
        } else {
            stats.fps           = (1.0 - alpha) * stats.fps + alpha * inst_fps;
            stats.bytes_per_sec = (1.0 - alpha) * stats.bytes_per_sec + alpha * inst_bps;
        }
    }
    stats.last_time = now;
}

std::string format_rate(double bytes_per_sec) {
    std::ostringstream oss;
    if (bytes_per_sec > 1024.0 * 1024.0) {
        oss << std::fixed << std::setprecision(1) << (bytes_per_sec / (1024.0 * 1024.0)) << " MB/s";
    } else if (bytes_per_sec > 1024.0) {
        oss << std::fixed << std::setprecision(1) << (bytes_per_sec / 1024.0) << " KB/s";
    } else {
        oss << std::fixed << std::setprecision(0) << bytes_per_sec << " B/s";
    }
    return oss.str();
}

struct Stream {
    std::string name;
    Subscriber<CompressedMessage> subscriber;
    StreamStats stats;
    cv::Mat latest_image;
    mutable std::mutex mutex;
    bool has_frame = false;

    explicit Stream(const std::string &topic_name) : name(topic_name), subscriber(topic_name) {}
};
}  // namespace

int main(int argc, char **argv) {
    std::signal(SIGINT, handle_signal);
    std::signal(SIGTERM, handle_signal);

    if (argc > 1) {
        domain_id = std::atoi(argv[1]);
    } else {
        domain_id = 10;  // Default domain ID
    }

    std::vector<std::string> topics;
    for (int i = 2; i < argc; ++i) {
        topics.emplace_back(argv[i]);
    }
    if (topics.empty()) {
        topics = {
            "igris_c/sensor/d435_color", "igris_c/sensor/d435_depth", "igris_c/sensor/eyes_stereo",
            "igris_c/sensor/left_hand",  "igris_c/sensor/right_hand",
        };
    }

    ChannelFactory::Instance()->Init(domain_id);
    if (!ChannelFactory::Instance()->IsInitialized()) {
        std::cerr << "Failed to initialize ChannelFactory" << std::endl;
        return 1;
    }

    IgrisC_Client client;
    client.Init();
    client.SetTimeout(10.0f);

    std::vector<std::unique_ptr<Stream>> streams;
    streams.reserve(topics.size());
    for (const auto &name : topics) {
        streams.emplace_back(std::make_unique<Stream>(name));
        cv::namedWindow(name, cv::WINDOW_NORMAL);
    }

    std::vector<status> stream_status(streams.size());
    for (size_t i = 0; i < streams.size(); ++i) {
        stream_status[i].name = streams[i]->name;

        if (!streams[i]->subscriber.init([stream = streams[i].get()](const CompressedMessage &msg) {
                const auto &bytes = msg.image_data();
                if (bytes.empty()) {
                    return;
                }

                cv::Mat raw(1, static_cast<int>(bytes.size()), CV_8UC1, const_cast<uint8_t *>(bytes.data()));
                cv::Mat img = cv::imdecode(raw, cv::IMREAD_UNCHANGED);
                if (img.empty()) {
                    return;
                }

                std::lock_guard<std::mutex> lock(stream->mutex);
                update_stats(stream->stats, bytes.size());
                stream->latest_image = img;
                stream->has_frame    = true;
            })) {
            std::cerr << "Failed to initialize subscriber for topic: " << streams[i]->name << std::endl;
            return 1;
        }
    }

    while (!g_stop.load()) {
        bool got_data = false;

        for (size_t i = 0; i < streams.size(); ++i) {
            auto &stream = *streams[i];
            cv::Mat frame;
            double fps           = 0.0;
            double bytes_per_sec = 0.0;
            {
                std::lock_guard<std::mutex> lock(stream.mutex);
                if (stream.has_frame) {
                    frame = stream.latest_image.clone();
                }
                fps           = stream.stats.fps;
                bytes_per_sec = stream.stats.bytes_per_sec;
            }

            if (!frame.empty()) {
                got_data = true;

                std::ostringstream oss;
                oss << std::fixed << std::setprecision(1) << fps << " FPS, " << format_rate(bytes_per_sec);
                cv::Scalar text_color = (fps <= 20.0 ? cv::Scalar(0, 0, 255) : cv::Scalar(0, 255, 0));
                cv::putText(frame, oss.str(), cv::Point(10, 30), cv::FONT_HERSHEY_SIMPLEX, 0.8, text_color, 2);
                stream_status[i].fps           = static_cast<float>(fps);
                stream_status[i].bytes_per_sec = static_cast<float>(bytes_per_sec);
                cv::imshow(stream.name, frame);
            }
        }

        std::cout << "\x1B[2J\x1B[H";
        std::cout << "==============================\n\n";
        std::cout << "Domain ID: " << domain_id << "\n";
        std::cout << "Stream Status:\n";
        for (const auto &s : stream_status) {
            bool low_fps            = (s.fps <= 20.0f);
            const char *color_start = low_fps ? "\x1B[31m" : "\x1B[0m";
            const char *color_end   = "\x1B[0m";
            std::cout << color_start << s.name << ": " << std::fixed << std::setprecision(1) << s.fps << " FPS, "
                      << format_rate(s.bytes_per_sec) << color_end << "\n";
        }
        std::cout << std::endl;
        std::cout << "Total bytes/sec: "
                  << format_rate(std::accumulate(stream_status.begin(), stream_status.end(), 0.0f,
                                                 [](float sum, const status &s) { return sum + s.bytes_per_sec; }))
                  << "\n";
        std::cout << "Avg FPS: " << std::fixed << std::setprecision(1)
                  << (stream_status.empty() ? 0.0f
                                            : std::accumulate(stream_status.begin(), stream_status.end(), 0.0f,
                                                              [](float sum, const status &s) { return sum + s.fps; }) /
                                                  stream_status.size())
                  << " FPS\n";
        std::cout << "Press 'q' or 'ESC' to quit.\n";
        std::cout << "==============================\n";
        std::cout << std::flush;

        int key = cv::waitKey(1);
        if (key == 27 || key == 'q' || key == 'Q') {
            break;
        }

        if (!got_data) {
            std::this_thread::sleep_for(std::chrono::milliseconds(2));
        }
    }

    for (const auto &stream : streams) {
        cv::destroyWindow(stream->name);
    }
    return 0;
}

File: examples/python/sensor_viewer.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import signal
import sys
import threading
import time
from dataclasses import dataclass, field

import cv2
import igris_c_sdk as igc_sdk
import numpy as np

DEFAULT_TOPICS = [
    "igris_c/sensor/d435_color",
    "igris_c/sensor/d435_depth",
    "igris_c/sensor/eyes_stereo",
    "igris_c/sensor/left_hand",
    "igris_c/sensor/right_hand",
]


def format_rate(bytes_per_sec: float) -> str:
    if bytes_per_sec > 1024.0 * 1024.0:
        return f"{bytes_per_sec / (1024.0 * 1024.0):.1f} MB/s"
    if bytes_per_sec > 1024.0:
        return f"{bytes_per_sec / 1024.0:.1f} KB/s"
    return f"{bytes_per_sec:.0f} B/s"


@dataclass
class StreamState:
    name: str
    fps: float = 0.0
    bytes_per_sec: float = 0.0
    last_time: float = 0.0
    last_payload: bytes | None = None
    sequence: int = 0
    decoded_sequence: int = -1
    frame: np.ndarray | None = None
    lock: threading.Lock = field(default_factory=threading.Lock)

    def update(self, payload: bytes) -> None:
        now = time.perf_counter()
        with self.lock:
            if self.last_time > 0.0:
                elapsed = now - self.last_time
                if elapsed > 0.0:
                    inst_fps = 1.0 / elapsed
                    inst_bps = len(payload) / elapsed
                    alpha = 0.15
                    if self.fps <= 0.0:
                        self.fps = inst_fps
                        self.bytes_per_sec = inst_bps
                    else:
                        self.fps = (1.0 - alpha) * self.fps + alpha * inst_fps
                        self.bytes_per_sec = (1.0 - alpha) * self.bytes_per_sec + alpha * inst_bps
            self.last_time = now
            self.last_payload = payload
            self.sequence += 1

    def decode_latest(self) -> np.ndarray | None:
        with self.lock:
            if self.last_payload is None or self.sequence == self.decoded_sequence:
                return self.frame
            payload = self.last_payload
            sequence = self.sequence

        raw = np.frombuffer(payload, dtype=np.uint8)
        frame = cv2.imdecode(raw, cv2.IMREAD_UNCHANGED)
        if frame is None or frame.size == 0:
            return None

        with self.lock:
            if self.sequence == sequence:
                self.frame = frame
                self.decoded_sequence = sequence
            return self.frame

    def snapshot(self) -> tuple[float, float]:
        with self.lock:
            return self.fps, self.bytes_per_sec


def print_status(streams: list[StreamState]) -> None:
    sys.stdout.write("\x1b[2J\x1b[H")
    sys.stdout.write("==============================\n\n")
    sys.stdout.write("Stream Status:\n")

    total_bps = 0.0
    total_fps = 0.0
    for stream in streams:
        fps, bps = stream.snapshot()
        total_bps += bps
        total_fps += fps
        color_start = "\x1b[31m" if fps <= 20.0 else "\x1b[0m"
        sys.stdout.write(f"{color_start}{stream.name}: {fps:.1f} FPS, {format_rate(bps)}\x1b[0m\n")

    avg_fps = total_fps / len(streams) if streams else 0.0
    sys.stdout.write("\n")
    sys.stdout.write(f"Total bytes/sec: {format_rate(total_bps)}\n")
    sys.stdout.write(f"Avg FPS: {avg_fps:.1f} FPS\n")
    sys.stdout.write("Press 'q' or 'ESC' to quit.\n")
    sys.stdout.write("==============================\n")
    sys.stdout.flush()


def build_arg_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="IGRIS compressed sensor stream viewer")
    parser.add_argument(
        "--domain-id",
        type=int,
        default=10,
        help="Cyclone DDS domain id (default: 10)",
    )
    parser.add_argument(
        "topics",
        nargs="*",
        help="Compressed image topics to subscribe to",
    )
    return parser


def create_subscribers(domain_id: int, streams: list[StreamState]) -> list[object]:
    channel = igc_sdk.ChannelFactory.Instance()
    channel.Init(domain_id)

    subscribers: list[object] = []
    for stream in streams:
        subscriber = igc_sdk.CompressedMessageSubscriber(stream.name)
        subscriber.init(lambda msg, stream=stream: stream.update(bytes(msg.image_data())))
        subscribers.append(subscriber)
    return subscribers


def main() -> int:
    args = build_arg_parser().parse_args()
    topics = args.topics or DEFAULT_TOPICS
    streams = [StreamState(name=topic) for topic in topics]

    stop_event = threading.Event()

    def handle_signal(_signum: int, _frame: object) -> None:
        stop_event.set()

    signal.signal(signal.SIGINT, handle_signal)
    signal.signal(signal.SIGTERM, handle_signal)

    subscribers = create_subscribers(args.domain_id, streams)

    for stream in streams:
        cv2.namedWindow(stream.name, cv2.WINDOW_NORMAL)

    last_status_print = 0.0
    try:
        while not stop_event.is_set():
            got_data = False

            for stream in streams:
                frame = stream.decode_latest()
                if frame is None:
                    continue

                got_data = True
                display = frame.copy()
                fps, bps = stream.snapshot()
                text = f"{fps:.1f} FPS, {format_rate(bps)}"
                color = (0, 0, 255) if fps <= 20.0 else (0, 255, 0)
                cv2.putText(
                    display,
                    text,
                    (10, 30),
                    cv2.FONT_HERSHEY_SIMPLEX,
                    0.8,
                    color,
                    2,
                )
                cv2.imshow(stream.name, display)

            now = time.perf_counter()
            if now - last_status_print >= 0.5:
                print_status(streams)
                last_status_print = now

            key = cv2.waitKey(1) & 0xFF
            if key in (27, ord("q"), ord("Q")):
                break

            if not got_data:
                time.sleep(0.002)
    finally:
        for subscriber in subscribers:
            subscriber.stop()
        for stream in streams:
            cv2.destroyWindow(stream.name)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
